sw/host/opentitanlib/src/rom_patch/patch.rs

```diff
@@ -0,0 +1,738 @@
+struct RomPatchRegionHjson {
+    match_addr: String,
+    locked: bool,
+    patch_code: Vec<String>,
+    sub_routines: Option<Vec<RomPatchSubRoutineHjson>>,
+}
+
+#[derive(Deserialize, Debug)]
+struct RomPatchHjson {
+    revision: u8,
+    base_addr: String,
+    regions: Vec<RomPatchRegionHjson>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct RomPatchPartitionHjson {
+    patches: Vec<RomPatchHjson>,
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+#[allow(dead_code)]
+#[repr(packed)]
+struct RomPatchMatchDescriptor {
+    match_base: u32,
+    remap_base: u32,
+}
+
+impl RomPatchMatchDescriptor {
+    fn from_hjson(remap_base: u32, region: &RomPatchRegionHjson) -> Result<Self> {
+        let match_addr = u32::from_str(&region.match_addr)?;
+
+        let size_dwords = (region.patch_code.len() as u32) & 0xf;
+        let locked = if region.locked { 1 << 31 } else { 0 };
+
+        let match_address = match_addr & 0x7ffffff;
+        let match_base = locked | (size_dwords << 27) | match_address;
+
+        Ok(RomPatchMatchDescriptor {
+            match_base,
+            remap_base,
+        })
+    }
+
+    fn match_addr(&self) -> u32 {
+        self.match_base & 0xffff
+    }
+
+    fn remap_addr(&self) -> u32 {
+        self.remap_base
+    }
+
+    fn locked(&self) -> bool {
+        (self.match_base >> 31) != 0
+    }
+
+    fn size(&self) -> u8 {
+        ((self.match_base >> 27) & 0xf) as u8
+    }
+
+    fn enabled(&self) -> bool {
+        self.match_base != 0 && self.remap_base != 0
+    }
+}
+
+impl fmt::Display for RomPatchMatchDescriptor {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "\tMatch address {:#04x}", self.match_addr())?;
+        writeln!(f, "\tRemap address {:#04x}", self.remap_addr())?;
+        write!(
+            f,
+            "\tPatch size {:?} Locked {:?} ",
+            self.size(),
+            self.locked()
+        )
+    }
+}
+
+#[derive(Default, Clone, Copy, Debug)]
+#[allow(dead_code)]
+#[repr(packed)]
+struct RomPatchHeader {
+    meta: u8,
+    size: u16,
+    revision: u8,
+}
+
+impl Read for RomPatchHeader {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let raw_header = self as *const _ as *const u8;
+        let header_size = mem::size_of::<RomPatchHeader>();
+
+        if buf.len() < header_size {
+            return Err(std::io::Error::from(std::io::ErrorKind::InvalidInput));
+        }
+
+        unsafe {
+            ptr::copy_nonoverlapping(raw_header, buf.as_mut_ptr(), header_size);
+        }
+
+        Ok(header_size)
+    }
+}
+
+impl fmt::Display for RomPatchHeader {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let unaligned = std::ptr::addr_of!(self.size);
+        let size = unsafe { std::ptr::read_unaligned(unaligned) };
+        write!(
+            f,
+            "\tMeta {:#x} Revision {:?} Size {:?} DWORDs",
+            self.meta, self.revision, size
+        )
```

Comment:
```suggestion
        write!(
            f,
            "\tMeta {:#x} Revision {:?} Size {:?} DWORDs",
            self.meta, self.revision, {self.size}
        )
```

Reply:
This part of the code has been reworked. We should not do unaligned accesses anymore, except in the unit tests.